pkg/reconciler/v1alpha1/serverlessservice/serverlessservice.go

```diff
@@ -0,0 +1,347 @@
+		return err
+	}
+	if !equality.Semantic.DeepEqual(sks.Status, original.Status) {
+		// Only update status if it changed.
+		if _, err = r.updateStatus(sks); err != nil {
+			r.Recorder.Eventf(sks, corev1.EventTypeWarning, "UpdateFailed", "Failed to update status: %v", err)
+			return err
+		}
+		r.Recorder.Eventf(sks, corev1.EventTypeNormal, "Updated", "Successfully updated ServerlessService %q", key)
+	}
+	return nil
+}
+
+func (r *reconciler) reconcile(ctx context.Context, sks *netv1alpha1.ServerlessService) error {
+	logger := logging.FromContext(ctx)
+	// Don't reconcile if we're being deleted.
+	if sks.GetDeletionTimestamp() != nil {
+		return nil
+	}
+
+	sks.SetDefaults(ctx)
+	sks.Status.InitializeConditions()
+
+	for i, fn := range []func(context.Context, *netv1alpha1.ServerlessService) error{
+		r.reconcilePrivateService,  // First make sure our data source is setup.
+		r.reconcilePublicService,   // Make sure the service is setup.
+		r.reconcilePublicEndpoints, // Now populate endpoints, if there are healthy ones.
+	} {
+		if err := fn(ctx, sks); err != nil {
+			logger.Debugw(fmt.Sprintf("%d: reconcile failed", i), zap.Error(err))
+			return err
+		}
+	}
+	sks.Status.ObservedGeneration = sks.Generation
+	return nil
+}
+
+func (r *reconciler) updateStatus(sks *netv1alpha1.ServerlessService) (*netv1alpha1.ServerlessService, error) {
+	original, err := r.sksLister.ServerlessServices(sks.Namespace).Get(sks.Name)
+	if err != nil {
+		return nil, err
+	}
+	if reflect.DeepEqual(original.Status, sks.Status) {
+		return original, nil
+	}
+	r.Logger.Debugf("StatusDiff: %s", cmp.Diff(original.Status, sks.Status))
+	original = original.DeepCopy()
+	original.Status = sks.Status
+	return r.ServingClientSet.NetworkingV1alpha1().ServerlessServices(sks.Namespace).UpdateStatus(original)
+}
+
+func (r *reconciler) reconcilePublicService(ctx context.Context, sks *netv1alpha1.ServerlessService) error {
+	logger := logging.FromContext(ctx)
+	sn := names.PublicService(sks)
+	srv, err := r.serviceLister.Services(sks.Namespace).Get(sn)
+	if errors.IsNotFound(err) {
+		logger.Infof("K8s service %q does not exist; creating.", sn)
+		// We've just created the service, so it has no endpoints.
+		sks.Status.MarkEndpointsUnknown("Creating")
+		srv = resources.MakePublicService(sks)
+		_, err := r.KubeClientSet.CoreV1().Services(sks.Namespace).Create(srv)
+		if err != nil {
+			logger.Errorw(fmt.Sprintf("Error creating K8s Service %s: ", sn), zap.Error(err))
+			return err
+		}
+		logger.Infof("Created K8s service: %q", sn)
+	} else if err != nil {
+		logger.Errorw(fmt.Sprintf("Error getting K8s Service %s: ", sn), zap.Error(err))
+		return err
+	} else if !metav1.IsControlledBy(srv, sks) {
+		sks.Status.MarkEndpointsNotOwned("Service", sn)
+		return fmt.Errorf("SKS: %q does not own Service: %q", sks.Name, sn)
+	} else {
+		tmpl := resources.MakePublicService(sks)
+		want := srv.DeepCopy()
+		want.Spec.Ports = tmpl.Spec.Ports
+		want.Spec.Selector = nil
```

Comment:
Do you not have to copy over the ClusterIP?

Reply:
No each service has its own. Also, k8s rejects changes to clusterIP, IIRC.

Reply:
I was reading the copy backwards, thanks.